pkg/controller/submariner/submariner_controller.go

```diff
@@ -367,19 +368,32 @@ func newRouteAgentDaemonSet(cr *submarinerv1alpha1.Submariner) *appsv1.DaemonSet
 	return routeAgentDaemonSet
 }
 
+const (
+	letters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+)
+
 //TODO: move to a method on the API definitions, as the example shown by the etcd operator here :
 //      https://github.com/coreos/etcd-operator/blob/8347d27afa18b6c76d4a8bb85ad56a2e60927018/pkg/apis/etcd/v1beta2/cluster.go#L185
 func setSubmarinerDefaults(submariner *submarinerv1alpha1.Submariner) {
 
 	spec := submariner.Spec
-	if spec.Repository == "" {
+	if len(spec.Repository) == 0 {
 		// An empty field is converted to the default upstream submariner repository where all images live
 		spec.Repository = "quay.io/submariner"
 	}
 
-	if spec.Version == "" {
+	if len(spec.Version) == 0 {
 		spec.Version = "0.0.2"
 	}
+
+	if len(spec.ClusterID) == 0 {
+		// Generate a random, 16-character cluster id
```

Comment:
Hmm, big question, if we do this, shouldn't we "write-back" the CR in the API?, otherwise, on next run we would generate a different clusterID because it's again empty.

Reply:
Good point, we need that anyway to be able to debug...

Reply:
This is now done in the reconciliation function.